core/src/main/java/com/linecorp/armeria/internal/server/annotation/FileAggregatedMultipart.java

```diff
@@ -0,0 +1,123 @@
+        this.params = params;
+        this.files = files;
+    }
+
+    ListMultimap<String, String> params() {
+        return params;
+    }
+
+    ListMultimap<String, Path> files() {
+        return files;
+    }
+
+    static CompletableFuture<FileAggregatedMultipart> aggregateMultipart(ServiceRequestContext ctx,
+                                                                         HttpRequest req) {
+        final Path multipartUploadsLocation = ctx.config().multipartUploadsLocation();
+        return Multipart.from(req).collect(bodyPart -> {
+            if (bodyPart.filename() != null) {
+                final ScheduledExecutorService blockingExecutorService =
+                        ctx.blockingTaskExecutor().withoutContext();
+                return resolveTmpFile(multipartUploadsLocation.resolve("incomplete"),
+                                      blockingExecutorService)
+                        .thenComposeAsync(
+                                path -> bodyPart
+                                        .writeTo(path)
+                                        .thenCompose(ignore -> moveFile(
+                                                path,
+                                                multipartUploadsLocation.resolve("complete"),
+                                                blockingExecutorService))
+                                        .thenApply(completePath -> Maps.<String, Object>immutableEntry(
+                                                bodyPart.name(), completePath)),
+                                ctx.eventLoop());
+            }
+            return bodyPart.aggregate()
+                           .thenApply(aggregatedBodyPart -> Maps.<String, Object>immutableEntry(
+                                   bodyPart.name(), aggregatedBodyPart.contentUtf8()));
+        }).thenApply(result -> {
+            final ImmutableListMultimap.Builder<String, String> params = ImmutableListMultimap.builder();
+            final ImmutableListMultimap.Builder<String, Path> files =
+                    ImmutableListMultimap.builder();
+            for (Entry<String, Object> entry : result) {
+                final Object value = entry.getValue();
+                if (value instanceof Path) {
+                    files.put(entry.getKey(), (Path) value);
+                } else {
+                    params.put(entry.getKey(), (String) value);
+                }
+            }
+            return new FileAggregatedMultipart(params.build(), files.build());
+        });
+    }
+
+    private static CompletableFuture<Path> moveFile(Path file, Path targetDirectory,
+                                                    ExecutorService blockingExecutorService) {
+        return CompletableFuture.supplyAsync(() -> {
+            try {
+                if (!Files.exists(targetDirectory)) {
+                    Files.createDirectories(targetDirectory);
+                }
```

Comment:
Perhaps we can remove the existence check because `createDirectories()` will do it anyway, i.e. we save syscalls:

```suggestion
                Files.createDirectories(targetDirectory);
```